pkg/tsdb/cloudwatch/response_parser.go

```diff
@@ -0,0 +1,155 @@
+package cloudwatch
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/cloudwatch"
+	"github.com/grafana/grafana/pkg/components/null"
+	"github.com/grafana/grafana/pkg/tsdb"
+)
+
+func (e *CloudWatchExecutor) parseResponse(metricDataOutputs []*cloudwatch.GetMetricDataOutput, queries map[string]*cloudWatchQuery) ([]*cloudwatchResponse, error) {
+	mdr := make(map[string]map[string]*cloudwatch.MetricDataResult)
+
+	for _, mdo := range metricDataOutputs {
+		requestExceededMaxLimit := false
+		for _, message := range mdo.Messages {
+			if *message.Code == "MaxMetricsExceeded" {
+				requestExceededMaxLimit = true
+			}
+		}
+
+		for _, r := range mdo.MetricDataResults {
+			if _, ok := mdr[*r.Id]; !ok {
+				mdr[*r.Id] = make(map[string]*cloudwatch.MetricDataResult)
+				mdr[*r.Id][*r.Label] = r
+			} else if _, ok := mdr[*r.Id][*r.Label]; !ok {
+				mdr[*r.Id][*r.Label] = r
+			} else {
+				mdr[*r.Id][*r.Label].Timestamps = append(mdr[*r.Id][*r.Label].Timestamps, r.Timestamps...)
+				mdr[*r.Id][*r.Label].Values = append(mdr[*r.Id][*r.Label].Values, r.Values...)
+			}
+			queries[*r.Id].RequestExceededMaxLimit = requestExceededMaxLimit
+		}
+	}
+
+	cloudWatchResponses := make([]*cloudwatchResponse, 0)
+	for id, lr := range mdr {
+		response := &cloudwatchResponse{}
+		series, err := parseGetMetricDataTimeSeries(lr, queries[id])
+		if err != nil {
+			return cloudWatchResponses, err
+		}
+
+		response.series = series
+		response.SearchExpression = queries[id].SearchExpression
+		response.RefId = queries[id].RefId
+		response.Id = queries[id].Id
+		response.RequestExceededMaxLimit = queries[id].RequestExceededMaxLimit
+
+		cloudWatchResponses = append(cloudWatchResponses, response)
+	}
+
+	return cloudWatchResponses, nil
+}
+
+func parseGetMetricDataTimeSeries(metricDataResults map[string]*cloudwatch.MetricDataResult, query *cloudWatchQuery) (*tsdb.TimeSeriesSlice, error) {
+	result := tsdb.TimeSeriesSlice{}
+	for label, metricDataResult := range metricDataResults {
+		if *metricDataResult.StatusCode != "Complete" {
+			return &result, fmt.Errorf("Part of query failed: %s", *metricDataResult.StatusCode)
+		}
+
+		for _, message := range metricDataResult.Messages {
+			if *message.Code == "ArithmeticError" {
+				return &result, &queryError{fmt.Errorf("ArithmeticError in query %s: %s", query.RefId, *message.Value), query.RefId}
+			}
+		}
+
+		series := tsdb.TimeSeries{
+			Tags:   map[string]string{},
+			Points: make([]tsdb.TimePoint, 0),
```

Comment:
Mixing make and non-make. Choose one, doesn't matter which one since the underlying representation is the same but in general in Grafana code base think make is usually used when there are a known length.